core/src/test/scala/org/broadinstitute/dsde/rawls/user/UserServiceSpec.scala

```diff
@@ -299,4 +304,427 @@ class UserServiceSpec extends AnyFlatSpecLike with TestDriverComponent with Mock
+        .getOrElse(fail("project not found")).billingAccount shouldEqual None
+    }
+  }
+
+  it should "not update the billing account for a billing project if the user does not have access to the billing project" in {
+    withMinimalTestDatabase { dataSource: SlickDataSource =>
+      val billingProject = minimalTestData.billingProject
+      val billingAccountName = RawlsBillingAccountName("billingAccounts/111111-111111-111111")
+      val newBillingAccountRequest = UpdateRawlsBillingAccountRequest(billingAccountName)
+
+      val mockSamDAO = mock[SamDAO](RETURNS_SMART_NULLS)
+      when(mockSamDAO.userHasAction(SamResourceTypeNames.billingProject, billingProject.projectName.value, SamBillingProjectActions.updateBillingAccount, userInfo)).thenReturn(Future.successful(false))
+
+      val mockGcsDAO = mock[GoogleServicesDAO](RETURNS_SMART_NULLS)
+      when(mockGcsDAO.testBillingAccountAccess(ArgumentMatchers.eq(billingAccountName), any[UserInfo])).thenReturn(Future.successful(true))
+
+      val userService = getUserService(dataSource, mockSamDAO, mockGcsDAO)
+
+      val actual = intercept[RawlsExceptionWithErrorReport] {
+        Await.result(userService.updateBillingProjectBillingAccount(billingProject.projectName, newBillingAccountRequest), Duration.Inf) shouldEqual()
+      }
+
+      actual.errorReport.statusCode.get shouldEqual StatusCodes.Forbidden
+
+      // billing account should remain unchanged
+      runAndWait(rawlsBillingProjectQuery.load(billingProject.projectName))
+        .getOrElse(fail("project not found")).billingAccount shouldEqual billingProject.billingAccount
+    }
+  }
+
+  it should "not update the billing account for a billing project if the user does not have access to the billing account" in {
+    withMinimalTestDatabase { dataSource: SlickDataSource =>
+      val billingProject = minimalTestData.billingProject
+      val billingAccountName = RawlsBillingAccountName("billingAccounts/111111-111111-111111")
+      val newBillingAccountRequest = UpdateRawlsBillingAccountRequest(billingAccountName)
+
+      val mockSamDAO = mock[SamDAO](RETURNS_SMART_NULLS)
+      when(mockSamDAO.userHasAction(SamResourceTypeNames.billingProject, billingProject.projectName.value, SamBillingProjectActions.updateBillingAccount, userInfo)).thenReturn(Future.successful(true))
+
+      val mockGcsDAO = mock[GoogleServicesDAO](RETURNS_SMART_NULLS)
+      when(mockGcsDAO.testBillingAccountAccess(ArgumentMatchers.eq(billingAccountName), any[UserInfo])).thenReturn(Future.successful(false))
+
+      val userService = getUserService(dataSource, mockSamDAO, mockGcsDAO)
+
+      val actual = intercept[RawlsExceptionWithErrorReport] {
+        Await.result(userService.updateBillingProjectBillingAccount(billingProject.projectName, newBillingAccountRequest), Duration.Inf) shouldEqual()
+      }
+
+      actual.errorReport.statusCode.get shouldEqual StatusCodes.BadRequest
+
+      // billing account should remain unchanged
```

Comment:
I also added these checks that the billing account in the database is what we expect it to be. This used to be covered by monitoring the calls to Google, but as I said above that doesn't belong here at this point since it's handled asynchronously.